torch/cuda/streams.py

```diff
@@ -119,90 +102,75 @@ def __repr__(self):
                 .format(self.device, self.cuda_stream))
 
 
-class EventHandle(ctypes.Structure):
-    IPC_HANDLE_SIZE = 64
-    _fields_ = [('reserved', ctypes.c_char * IPC_HANDLE_SIZE)]
-
-
-class Event(object):
-    r"""Wrapper around CUDA event.
+class Event(torch._C._CudaEventBase):
+    r"""Wrapper around CUDA event. Every event is associated with a device index, which
+    is the device where the event is lazily created when first recorded. However, if
+    constructed from a handle or ipc_handle() is called before it is ever recorded, the
+    device index will be the sames as the device of ``torch.cuda.current_stream()``.
 
     Arguments:
-        enable_timing (bool): indicates if the event should measure time
+        enable_timing (bool, optional): indicates if the event should measure time
             (default: ``False``)
-        blocking (bool): if ``True``, :meth:`wait` will be blocking (default: ``False``)
+        blocking (bool, optional): if ``True``, :meth:`wait` will be blocking (default: ``False``)
         interprocess (bool): if ``True``, the event can be shared between processes
             (default: ``False``)
+        _handle (bytes-like object, optional): acquired by calling Event.ipc_handle()
```

Comment:
I don't think we should document `_handle`. It's intended as a private API. The preferred way to pass an even to another process is to pass the event itself (not to call `ipc_handle()` and `Event(_handle=...)`)

Reply:
I thought about this a bit, and am a little inclined to think that using handle might actually be better in some cases. Suppose we have a parent process spawning a bunch of children process, with each child working on a different device. We have at least two options to synchronize parent and children:

Option 1. Create all events in the parent process, and pass an event (associated with the correct device) to each child. In this case, the parent process need to do many ctx switching, and will need to use a loop synchronize with all children. Moreover, if children also need to synchronize between each other, every child need to keep all other children's event object, and will also need to use a loop to synchronize.

Option 2. Create an event on master, and then pass the same handle to each child. In this case, every process has only a single event, and looks more concise. (actually, let me add a test for creating two events from the same handle).

I don't have a strong preference, and open to any discussion.

Reply:
You can implement both cases in user code without ever calling the `Event(_handle=...)` constructor or explicitly calling `ipc_handle()`.

Something like:

```python
event = torch.cuda.Event()
p1 = mp.Process(target=target1, args=(event,))
p2 = mp.Process(target=target1, args=(event,))
p3 = mp.Process(target=target1, args=(event,))
p1.start(); p2.start(); p3.start()
```

It's better to pass the `event` instead of the handle because then we can encode the device and ensure that it's opened on the correct device.

Reply:
I might miss something. I thought if p1, p2, and p3 operate on different devices, they cannot use that event to record, no?

Reply:
I think we have an unverified assumption regarding whether an event can be reconstructed from handle on any device. Let me run some more tests.